src/System.Net.Http/ref/System.Net.Http.cs

```diff
@@ -191,6 +191,7 @@ public partial class HttpResponseMessage : System.IDisposable
         public HttpResponseMessage(System.Net.HttpStatusCode statusCode) { }
         public System.Net.Http.HttpContent Content { get { throw null; } set { } }
         public System.Net.Http.Headers.HttpResponseHeaders Headers { get { throw null; } }
+        public System.Net.Http.Headers.HttpResponseHeaders TrailingHeaders { get { throw null; } set { } }
```

Comment:
If you have to have a setter, then it should not be a public property.  It should be internal only.

Reply:
@davidsh beware of lockin. It should be possible for others to implement transports that use trailers without access to internals fields. E.g. we have a handler implementation [here](https://github.com/aspnet/AspNetCore/blob/67a6f08da94910b1a0e8fb45a30235d1ad16d6df/src/Hosting/TestHost/src/ClientHandler.cs) that's a full transport (in memory, adapted over HttpContext). HttpClient and family have always been a fully open and re-usable object model. Fields and collections all need to be publicly mutable in some form.

Reply:
>Fields and collections all need to be publicly mutable in some form.

The existing 'Headers' property here doesn't have a 'setter'.  How are you handling that?  It seems inconsistent to add a public property setter for this collection property when all the others don't have one.

Reply:
The other headers don't need a setter because they always come pre-initialized with mutable collections. Defaulting to null for TrailingHeaders means you also need a way to initialize it.

Reply:
And if I had to choose between A) addressing the ambiguity of an empty collection or B) maintaining a reusable object model and being able to implement trailers in other handlers, B is far more important.

Reply:
Seems there are two goals here:
1. Allow arbitrary handlers to use this object model.
2. Not force the 99.9% case of no trailing headers to incur a perf penalty of even more allocation than this object model already incurs.

And that suggests we either:
A. Have a get/set TrailingHeaders property, with null to indicate we don't yet know whether there are trailers, and non-null to indicate we do, then using a read-only empty collection to avoid having to instantiate one for someone to check whether there are any.
B. Have a gettable TrailingHeaders property and a get/set TrailingHeadersStatus property that returns one of NotYetAvailable (the response is still being processed), None (the response was processed and there were no trailing headers), and Available (the response was processed and there were trailing headers), such that you only access TrailingHeaders if Available, and you know whether or not you can access it yet with NotYetAvailable… then the object model would instantiate the TrailingHeaders collection lazily as with all other collections, the handler would only touch it when headers needed to be filled in, and a consumer who cared about perf would only touch it when Available (a consumer who didn't care about perf would touch it when != NotYetAvailable).

Reply:
If I had to choose between A or B, I would choose A.  Adding two properties plus an enum type for this feature seems like too much API clutter.  This feature is not going to be used by the majority of HttpClient API consumers.

Reply:
(B) could also be a `bool? HasTrailingHeaders { get; set; }` and `Collection TrailingHeaders { get; }`, getting rid of the enum in favor of a nullable bool.

Reply:
Where TrailingHeaders is lazy initialized?

Reply:
I wonder about the properties getting out of sync. It seems like HasTrailingHeaders could automatically be set if someone mutated TrailingHeaders. `get => _tralers?.Count > 0`;

Reply:
I also prefer A (current approach) to reduce complexity, it's just cleaner - the number of words in the description by Stephen is a good indicator :) 

The setter will remain to be `public` because as Chris has mentioned, maintaining a reusable object model is more important.

Reply:
That's not quite enough, HttpResponseHeaders's constructor is also internal. TrailingHeaders either needs to be lazy initialized like the others or HttpResponseHeaders needs a public constructor. That's not a new request, others have complained about not being able to new up the header collections.

Reply:
>That's not a new request, others have complained about not being able to new up the header collections.

If we're going to add API for public ctor of HttpResponseHeaders, then we should it also for HttpRequestHeaders and HttpContentHeaders.